ethportal-peertest/src/jsonrpc.rs

```diff
@@ -3,183 +3,266 @@ use std::io::prelude::*;
+        Test::new(
+            JsonRpcRequest {
+                method: "portal_historyFindNodes".to_string(),
+                id: 7,
+                params: Params::Array(vec![
+                    Value::String(peertest.bootnode.enr.to_base64()),
+                    Value::Array(vec![json!(256u16)]),
+                ]),
+            },
+            validate_portal_history_find_nodes,
+        ),
+        // Test FindNodes with 0 distance -> Peer enr
+        Test::new(
+            JsonRpcRequest {
                 method: "portal_historyFindNodes".to_string(),
                 id: 8,
                 params: Params::Array(vec![
-                    Value::String(bootnode.clone()),
-                    Value::Array(vec![json!(256)]),
+                    Value::String(peertest.bootnode.enr.to_base64()),
+                    Value::Array(vec![json!(0u16)]),
                 ]),
             },
-            JsonRpcEndpoint {
+            validate_portal_history_find_nodes_zero_distance,
+        ),
+        // Test FindNodes with 256 distance -> Routing Table enrs
+        Test::new(
+            JsonRpcRequest {
                 method: "portal_stateFindNodes".to_string(),
                 id: 9,
                 params: Params::Array(vec![
-                    Value::String(bootnode),
-                    Value::Array(vec![json!(256)]),
+                    Value::String(peertest.bootnode.enr.to_base64()),
+                    Value::Array(vec![json!(256u16)]),
                 ]),
             },
-        ]
-    }
+            validate_portal_state_find_nodes,
+        ),
+        // Test FindNodes with 0 distance -> Peer enr
+        Test::new(
+            JsonRpcRequest {
+                method: "portal_stateFindNodes".to_string(),
+                id: 10,
+                params: Params::Array(vec![
+                    Value::String(peertest.bootnode.enr.to_base64()),
+                    Value::Array(vec![json!(0u16)]),
+                ]),
+            },
+            validate_portal_state_find_nodes_zero_distance,
+        ),
+    ]
+}
 
-    pub fn to_jsonrpc(&self) -> String {
-        match self.params {
-            Params::None => format!(
-                r#"
-                {{
-                    "jsonrpc":"2.0",
-                    "id": {},
-                    "method": "{}"
-                }}"#,
-                self.id, self.method
-            ),
-            _ => format!(
-                r#"
-                {{
-                    "jsonrpc":"2.0",
-                    "id": {},
-                    "method": "{}",
-                    "params":{}
-                }}"#,
-                self.id,
-                self.method,
-                serde_json::to_string(&self.params).unwrap()
-            ),
-        }
-    }
+fn validate_web3_client_version(val: &Value, _peertest: &Peertest) {
+    assert_eq!(val.as_str().unwrap(), "trin v0.1.0");
+}
```

Comment:
So I know it was already like this, but since we're refactoring... I find it hard to quickly understand the test when the setup and validation are so separated. I am in favor of figuring out a way to combine them. Maybe even just inlining this function as an anonymous argument to the `Test`?

Reply:
Definitely agreed. I spent a fair amount of time trying to find a better solution, but I just kept on running into problems with the rust compiler. My approach was to inline the `validate_*` code as a closure on `Test`. The problem is that rust's compiler treats every closure as an individual type, regardless of what it accepts/returns. The only way around this that I could find was by wrapping the validation logic in a function with a defined signature. Is this similar to what you had in mind by "inlining this function as an anonymous argument"?

Reply:
Hm, it sounds like you maybe tried the thing that I was hoping would work, something like:
```rust
        Test::new(
            JsonRpcRequest {
                method: "portal_statePing".to_string(),
                id: 5,
                params: Params::Array(vec![
                    Value::String(peertest.bootnode.enr.to_base64()),
                    Value::String(DATA_RADIUS.to_owned()),
                ]),
            },
            |result: &Value, _peertest| {
                assert_eq!(
                    result.get("dataRadius").unwrap().as_str().unwrap(),
                    DATA_RADIUS
                );  
                assert_eq!(
                    result.get("enrSeq").unwrap().as_str().unwrap(),
                    ENR_SEQ.to_string()
                );  
            },
        ),
```

The main goal is to get the lines nearby each other that are defining the request made and the validation performed.

Reply:
Yup, exactly that. Unfortunately it's a nogo with the compiler. But I'll definitely keep this goal in mind as I add tests and hopefully figure out something that achieves this goal